src/clarity-icons/extend-clarity-icons.ts

```diff
@@ -0,0 +1,98 @@
+import { IconTemplate } from "./icon-template";
+import { IconAlias } from "./icon-alias";
+import { SVG_IMG_ICON_TEMPLATES } from "./svg-img-icon-templates";
+import { SVG_ICON_TEMPLATES } from "./svg-icon-templates";
+
+let EXTENDED_ICON_TEMPLATES: IconTemplate = {};
+
+export class UserIcons {
+
+    private validateName(name: string): boolean {
+
+        return name.length !== 0;
+
+    }
+
+    private validateTemplate(template: string): boolean {
+
+        return template.length !== 0;
+
+    }
+
+    getExtendedShapes(): IconTemplate {
+
+        return EXTENDED_ICON_TEMPLATES;
+    }
+
+
+    private setIconTemplate(shapeName: string, shapeTemplate: string): void {
+
+        if (this.validateName(shapeName) && this.validateTemplate(shapeTemplate)) {
+
+            EXTENDED_ICON_TEMPLATES[ shapeName ] = shapeTemplate;
+        }
+
+    }
+
+    addIcons(icons: IconTemplate): void {
+
+        for (let shapeName in icons) {
+
+            if (icons.hasOwnProperty(shapeName)) {
+
+                this.setIconTemplate(shapeName, icons[ shapeName ]);
+            }
+        }
+
+    }
+
+    private setIconAliases(templates: IconTemplate, shapeName: string, aliasNames: string[]): void {
+        for (let aliasName of aliasNames) {
+
+            Object.defineProperty(templates, aliasName, {
+                get: function () {
+                    return templates[ shapeName ];
+                },
+                enumerable: true,
+                configurable: true
+            });
+        }
+    }
+
+    addAliases(aliases: IconAlias): void {
+
+        for (let shapeName in aliases) {
+
+            if (aliases.hasOwnProperty(shapeName)) {
+
+                if (SVG_ICON_TEMPLATES.hasOwnProperty(shapeName)) {
```

Comment:
Both `SVG_ICON_TEMPLATES` AND `EXTENDED_ICON_TEMPLATES` are not accessible to users. They can be mutable only through `ClarityIcons` API methods.